src/components/inputs/date-time-input/date-time-input.js

```diff
@@ -1,349 +1,837 @@
+                />
+                <div className={styles.month}>{props.label}</div>
+                <SecondaryIconButton
+                  label="next month"
+                  onClick={() => {
+                    setMonth(
+                      moment(month)
+                        .add(1, 'month')
+                        .toDate()
+                    );
+                  }}
+                  icon={<AngleRightIcon size="medium" />}
+                />
+              </div>
+            </Heading>
+            <div className={styles.daysHeaderContainer}>
+              <div className={styles.daysHeader}>
+                {days.map(day => (
+                  <span key={day} className={styles.daysHeaderItem}>
+                    {day}
+                  </span>
+                ))}
+              </div>
+            </div>
+            <div className={styles.daysContainerContainer}>
+              <div className={styles.daysContainer}>{props.children}</div>
+            </div>
+          </div>
+        );
+      }}
+    </CalendarConnector.Consumer>
+  );
+});
+Group.displayName = 'Group';
+
+const Option = props => (
+  <CalendarConnector.Consumer>
+    {({ locale }) => {
+      if (props.data.display === 'calendar') {
+        const today = new Date();
+        const optionStyles = props.getStyles('option', {
+          ...props,
+          today: props.data.date.isSame(today, 'day'),
+        });
+        // Indent the first day of the month (date() === 1) in so that it starts
+        // at the appropriate position.
+        // Further respect the start of the week depending on the locale to
+        // adjust the indentation.
+        if (props.data.date.date() === 1) {
+          const firstDayOfWeek = moment.localeData(locale).firstDayOfWeek();
+          const indentBy = props.data.date.day() - firstDayOfWeek;
+          if (indentBy) {
+            optionStyles.marginLeft = `${indentBy * 14 + 1}%`;
+          }
+        }
+
+        return (
+          <span {...props.innerProps} style={optionStyles} ref={props.innerRef}>
+            {props.data.date.format('D')}
+          </span>
+        );
+      }
+      return <SelectComponents.Option {...props} />;
+    }}
+  </CalendarConnector.Consumer>
+);
+Option.displayName = 'Option';
+
+// This component is used so that we can enhance the keyboard navigation
+class SelectContainer extends Component {
+  static displayName = 'SelectContainer';
+  render() {
+    return (
+      <CalendarConnector.Consumer>
+        {({ selectRef, month, setMonth }) => (
+          <SelectComponents.SelectContainer
+            {...this.props}
+            innerProps={{
+              ...this.props.innerProps,
+              // Let react-select handle the keyDown event unless the menu
+              // is already open. When the menu is open, we want enhance
+              // the UX by customizing the keyboard usage.
+              onKeyDown: event => {
```

Comment:
In theory it could be reused (maybe). There are slight nuances in keyboard usage between the different components, so the handlers vary a little bit. It could be that it is the same handler at the moment, but the they could vary if we add more functionality - they are not guaranteed to always stay the same. So I would rather duplicate them.

Another thing which would bother me about extracting the function is that it is using the context. It would feel a bit weird to extract a non-pure function which accesses the context.